In [320]:
#Importando pacotes
import pandas as pd
import numpy as np

In [321]:
#Importando a planilha com os dados como DataFrame
dados = pd.read_excel('dados_olx10.xlsx')

In [322]:
#Observando a estrutura do DataFrame
dados.head()


title       price  locations  \
0  Arsenal Imóveis-Casa e vila, para aluguel com ...      R$ 500        NaN   
1  Arsenal Imóveis-Casa de vila para aluguel  com...      R$ 400        NaN   
2  GJB IMÓVEIS ALUGA APT NO GRANG HYATT BARRA 1 S...   R$ 16.000        NaN   
3  Arsenal Imóveis vende- Casa duplex com 1 Suíte...  R$ 210.000        NaN   
4  Asset Imóveis vende apartamento com 1 dormitór...  R$ 400.000        NaN   

     neighbourhood    municipality      category rooms bathrooms  
0          Tribobó     São Gonçalo  Apartamentos     1         1  
1          Tribobó     São Gonçalo         Casas     1         1  
2  Barra da Tijuca  Rio de Janeiro  Apartamentos     1         2  
3          Arsenal     São Gonçalo         Casas     2         1  
4           Icaraí         Niterói  Apartamentos     1         1

In [323]:
dados.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 27445 entries, 0 to 27444
Data columns (total 8 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   title          24950 non-null  object 
 1   price          24889 non-null  object 
 2   locations      0 non-null      float64
 3   neighbourhood  24875 non-null  object 
 4   municipality   24950 non-null  object 
 5   category       24950 non-null  object 
 6   rooms          20122 non-null  object 
 7   bathrooms      20102 non-null  object 
dtypes: float64(1), object(7)
memory usage: 1.7+ MB


In [324]:
#Quantidade de linhas do DataFrame
print(f'O DataFrame possui: {dados.shape[0]} linhas')
#Quantidade de colunas do DataFrame
print(f'O DataFrame possui: {dados.shape[1]} colunas')

O DataFrame possui: 27445 linhas
O DataFrame possui: 8 colunas


Como podemos ver, possuem mais linhas do que valores de fato, isso se deve as linhas completamente em branco.

In [325]:
#Primeiramente serão removidas as linhas em branco.
#Para garantir que o dropna vai remover todas as linhas em branco, vamos substituir as strings vazias por NaN
dados_sem_linha_em_branco = dados.replace('', np.nan)
dados_sem_linha_em_branco = dados.dropna(how='all')

In [326]:
#Vamos verificar se de fato alguma linha foi removida
print(f'Foram removidas {dados.shape[0]- dados_sem_linha_em_branco.shape[0]} linhas')

Foram removidas 2495 linhas


In [327]:
#Como vimos acima, 2495 linhas em branco foram removidas do DataFrame.
#Para não precisar utilizar um nome tão grande para o nosso DataFrame, vamos voltar a chamá-lo de dados.
dados = dados_sem_linha_em_branco.copy()

In [328]:
dados.info()

<class 'pandas.core.frame.DataFrame'>
Index: 24950 entries, 0 to 27444
Data columns (total 8 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   title          24950 non-null  object 
 1   price          24889 non-null  object 
 2   locations      0 non-null      float64
 3   neighbourhood  24875 non-null  object 
 4   municipality   24950 non-null  object 
 5   category       24950 non-null  object 
 6   rooms          20122 non-null  object 
 7   bathrooms      20102 non-null  object 
dtypes: float64(1), object(7)
memory usage: 1.7+ MB


In [329]:
dados = dados.reset_index(drop = True)

In [330]:
dados.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 24950 entries, 0 to 24949
Data columns (total 8 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   title          24950 non-null  object 
 1   price          24889 non-null  object 
 2   locations      0 non-null      float64
 3   neighbourhood  24875 non-null  object 
 4   municipality   24950 non-null  object 
 5   category       24950 non-null  object 
 6   rooms          20122 non-null  object 
 7   bathrooms      20102 non-null  object 
dtypes: float64(1), object(7)
memory usage: 1.5+ MB


In [331]:
#Vamos trabalhar apenas com o municipio do Rio de Janeiro, então precisamos filtrar
dados = dados[dados['municipality'] == 'Rio de Janeiro']
print(f'Fora removidas: {24950-dados.shape[0]} restando: {dados.shape[0]}')

Fora removidas: 9021 restando: 15929


In [332]:
#Vamos remover algumas colunas desnecessárias do DataFrame.
#Como a coluna 'locations' possui apenas entradas NaN, ela vai ser removida.
#Como todos os dados são referentes a cidade do Rio de Janeiro. A coluna que identifica  a cidade será removida.
#Sendo ela: 'municipality'.
dados = dados.drop(columns=['locations','municipality'])

In [333]:
#Agora o DataFrame ficou da seguinte forma
print(dados.head())

                                               title       price  \
2  GJB IMÓVEIS ALUGA APT NO GRANG HYATT BARRA 1 S...   R$ 16.000   
5  Privilegio Imóveis vende: Excelente apartament...  R$ 870.000   
6  Privilegio Imóveis vende: Excelente apartament...  R$ 870.000   
7  GJB IMÓVEIS VENDE LINDO  APT  RECREIO SALA AMP...  R$ 690.000   
8  GJB IMÓVEIS VENDE  LINDO APT NO RECREIO ,  3 q...  R$ 870.000   

              neighbourhood      category rooms bathrooms  
2           Barra da Tijuca  Apartamentos     1         2  
5                  Botafogo  Apartamentos     1         1  
6                  Botafogo  Apartamentos     1         1  
7  Recreio dos Bandeirantes  Apartamentos     3         2  
8  Recreio dos Bandeirantes  Apartamentos     3         3  


Agora serão removidas as linhas que possuem todas as colunas iguais exceto o 'price', pois se referem ao mesmo produto. O anunciantes costumam repostar o mesmo produto na plataforma e apenas mudar o preço. Será mantida a ultima informação por ser a mais atualizada.

In [334]:
#Removendo linhas com a mesma informação exceto o preço
#Obtendo lista com o nome das colunas 
colunas =  list(dados.columns)
#Removendo o preço desta lista
colunas.remove('price')
#Removendo as linhas com a mesma informação e mantendo a mais atual.
dados_sem_duplicadas = dados.drop_duplicates(subset=colunas, keep='last') 

In [335]:
#Número de linhas com as duplicadas
print(f'Número de linhas do DataFrame com as duplicadas: {dados.shape[0]}')
#Número de linhas sem as duplicadas
print(f'Número de linhas do DataFrame após remover as duplicadas: {dados_sem_duplicadas.shape[0]}')
#Número de linhas removidas
numero_de_linhas_removidas = dados.shape[0]-dados_sem_duplicadas.shape[0]
print(f'foram removidas {numero_de_linhas_removidas} linhas')

Número de linhas do DataFrame com as duplicadas: 15929
Número de linhas do DataFrame após remover as duplicadas: 13336
foram removidas 2593 linhas


In [336]:
#Para não ficar com um DataFrame que possui nome grande. Passaremos a chamar o 'dados_sem_duplicadas' apenas de dados
dados = dados_sem_duplicadas.copy()

In [337]:
dados = dados.reset_index(drop = True)
dados.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 13336 entries, 0 to 13335
Data columns (total 6 columns):
 #   Column         Non-Null Count  Dtype 
---  ------         --------------  ----- 
 0   title          13336 non-null  object
 1   price          13308 non-null  object
 2   neighbourhood  13336 non-null  object
 3   category       13336 non-null  object
 4   rooms          11507 non-null  object
 5   bathrooms      11497 non-null  object
dtypes: object(6)
memory usage: 625.3+ KB


In [338]:
#Como temos inicialmente o objetivo de ajustar um modelo de regressão, a coluna 'title' não será útil por não possuir um padrão, logo será removida.
dados = dados.drop(columns='title')
print(dados.head())

        price             neighbourhood      category rooms bathrooms
0  R$ 870.000                  Botafogo  Apartamentos     1         1
1  R$ 670.000               Jacarepaguá  Apartamentos     2         2
2  R$ 690.000  Recreio dos Bandeirantes  Apartamentos     2         2
3  R$ 500.000                Copacabana  Apartamentos     1         2
4   R$ 22.000                Santa Cruz         Casas     2         1


In [339]:
#A coluna neighbourhood possui muitas categorias, e isso é um problema na hora de ajustar um modelo. 
#Para resolver este problema será feita uma nova categorização da variável.
bairros = dados['neighbourhood'].unique()
print(bairros)


['Botafogo' 'Jacarepaguá' 'Recreio dos Bandeirantes' 'Copacabana'
 'Santa Cruz' 'Parque Anchieta' 'Campo Grande' 'Santíssimo' 'Guaratiba'
 'Freguesia (Jacarepaguá)' 'Laranjeiras' 'Ipanema' 'Flamengo' 'Pechincha'
 'Tijuca' 'Campinho' 'Leblon' 'São Conrado' 'Barra da Tijuca' 'Taquara'
 'Vidigal' 'Andaraí' 'São Cristóvão' 'Anchieta' 'Humaitá' 'Anil' 'Gamboa'
 'Maracanã' 'Bangu' 'Praça Seca' 'Cachambi' 'Marechal Hermes' 'Sepetiba'
 'Centro' 'Paciência' 'Itanhangá' 'Vargem Grande' 'Cosme Velho'
 'Vicente de Carvalho' 'Engenho de Dentro' 'Curicica' 'Rio Comprido'
 'Jardim Guanabara' 'Oswaldo Cruz' 'Irajá' 'Cidade Nova' 'Tomás Coelho'
 'Urca' 'Alto da Boa Vista' 'Jardim Botânico' 'Grajaú' 'Glória' 'Piedade'
 'Santa Teresa' 'Méier' 'Madureira' 'Gávea' 'Pedra de Guaratiba'
 'Vila Isabel' 'Magalhães Bastos' 'Senador Camará' 'Catete' 'Olaria'
 'Vargem Pequena' 'Braz de Pina' 'Maria da Graça' 'Realengo' 'Encantado'
 'Senador Vasconcelos' 'Saúde' 'Engenheiro Leal' 'Cascadura'
 'Rocha Miranda' 'Leme

In [340]:
print(f'A coluna neighbourhood possui: {len(bairros)} entradas distintas ')

A coluna neighbourhood possui: 145 entradas distintas 


Como possuem 145 bairros, seria muito trabalhoso fazer isso manualmente. Então será importada uma tabela com as informações dos bairros e a região ao qual eles pertencem que são 4. A coluna/variável com a região será utilizada no lugar do bairro na hora de fazer o modelo de regressão.

In [341]:
#Importando a tabela da wikipedia
url = 'https://pt.wikipedia.org/wiki/Lista_de_bairros_do_Rio_de_Janeiro_por_%C3%8Dndice_de_Desenvolvimento_Social'
html = pd.read_html(url, match='Bairro')
print(html)

[      Nº          Bairro              R.A      Região      IDS
0      1           Lagoa            Lagoa    Zona Sul    0.854
1      2          Leblon            Lagoa    Zona Sul  809.000
2      3         Ipanema            Lagoa    Zona Sul  801.000
3      4         Humaitá         Botafogo    Zona Sul  798.000
4      5            Urca         Botafogo    Zona Sul  795.000
..   ...             ...              ...         ...      ...
153  154       Guaratiba        Guaratiba  Zona Oeste  433.000
154  155  Vargem Pequena  Barra da Tijuca  Zona Oeste  425.000
155  156   Vargem Grande  Barra da Tijuca  Zona Oeste  408.000
156  157         Camorim      Jacarepaguá  Zona Oeste  369.000
157  158         Grumari  Barra da Tijuca  Zona Oeste  277.000

[158 rows x 5 columns]]


In [342]:
#Passando o conteúdo do html para o dataframe dados_bairros
type(html)
dados_bairros=html[0]
type(dados_bairros)
print(dados_bairros)

      Nº          Bairro              R.A      Região      IDS
0      1           Lagoa            Lagoa    Zona Sul    0.854
1      2          Leblon            Lagoa    Zona Sul  809.000
2      3         Ipanema            Lagoa    Zona Sul  801.000
3      4         Humaitá         Botafogo    Zona Sul  798.000
4      5            Urca         Botafogo    Zona Sul  795.000
..   ...             ...              ...         ...      ...
153  154       Guaratiba        Guaratiba  Zona Oeste  433.000
154  155  Vargem Pequena  Barra da Tijuca  Zona Oeste  425.000
155  156   Vargem Grande  Barra da Tijuca  Zona Oeste  408.000
156  157         Camorim      Jacarepaguá  Zona Oeste  369.000
157  158         Grumari  Barra da Tijuca  Zona Oeste  277.000

[158 rows x 5 columns]


In [343]:
#Salvando o dados_bairros em um planilha para não precisar fazer o scrapping sempre
dados_bairros.to_excel('dados_bairros_rj.xlsx', index=False)

In [344]:
#importando a planilha para ter certeza que funcionou
dados_bairros = pd.read_excel('dados_bairros_rj.xlsx')

In [345]:
dados_bairros.head()

Nº   Bairro       R.A    Região      IDS
0   1    Lagoa     Lagoa  Zona Sul    0.854
1   2   Leblon     Lagoa  Zona Sul  809.000
2   3  Ipanema     Lagoa  Zona Sul  801.000
3   4  Humaitá  Botafogo  Zona Sul  798.000
4   5     Urca  Botafogo  Zona Sul  795.000

In [346]:
#Renomeando as colunas do df inicial para que seja feito posteriormente uma junção dos dois dataframes
dados = dados.rename(columns={'price':'Preço','category':'Categoria','rooms':'Quartos','bathrooms':'Banheiros','neighbourhood':'Bairro'})

In [347]:
#Novo nome das colunas
dados.head()

Preço                    Bairro     Categoria Quartos Banheiros
0  R$ 870.000                  Botafogo  Apartamentos       1         1
1  R$ 670.000               Jacarepaguá  Apartamentos       2         2
2  R$ 690.000  Recreio dos Bandeirantes  Apartamentos       2         2
3  R$ 500.000                Copacabana  Apartamentos       1         2
4   R$ 22.000                Santa Cruz         Casas       2         1

In [348]:
#Merge dos dois dataframes com a intenção de obter a coluna/variável 'Região'
df_merged = pd.merge(left=dados, right=dados_bairros, how="left", on="Bairro")
print(df_merged.head())

        Preço                    Bairro     Categoria Quartos Banheiros  \
0  R$ 870.000                  Botafogo  Apartamentos       1         1   
1  R$ 670.000               Jacarepaguá  Apartamentos       2         2   
2  R$ 690.000  Recreio dos Bandeirantes  Apartamentos       2         2   
3  R$ 500.000                Copacabana  Apartamentos       1         2   
4   R$ 22.000                Santa Cruz         Casas       2         1   

      Nº              R.A      Região    IDS  
0   16.0         Botafogo    Zona Sul  743.0  
1  148.0      Jacarepaguá  Zona Oeste  476.0  
2   56.0  Barra da Tijuca  Zona Oeste  612.0  
3   14.0       Copacabana    Zona Sul  753.0  
4  147.0       Santa Cruz  Zona Oeste  476.0  


In [349]:
df_merged.head()

Preço                    Bairro     Categoria Quartos Banheiros  \
0  R$ 870.000                  Botafogo  Apartamentos       1         1   
1  R$ 670.000               Jacarepaguá  Apartamentos       2         2   
2  R$ 690.000  Recreio dos Bandeirantes  Apartamentos       2         2   
3  R$ 500.000                Copacabana  Apartamentos       1         2   
4   R$ 22.000                Santa Cruz         Casas       2         1   

      Nº              R.A      Região    IDS  
0   16.0         Botafogo    Zona Sul  743.0  
1  148.0      Jacarepaguá  Zona Oeste  476.0  
2   56.0  Barra da Tijuca  Zona Oeste  612.0  
3   14.0       Copacabana    Zona Sul  753.0  
4  147.0       Santa Cruz  Zona Oeste  476.0

In [350]:
#Checkando se realmente a coluna região possui apenas 4 entradas únicas (Zona Central,Zona Norte, Zona Oeste e Zona Sul)
print(len(df_merged['Região'].unique()))
print(df_merged['Região'].unique())

5
['Zona Sul' 'Zona Oeste' 'Zona Norte' 'Zona Central' nan]


Como podemos ver, nem todos os nomes de bairros estavam contidos no dados_bairros.
Será necessário ver os bairros que não foram classificados em uma região e classificar manualmente.

In [351]:
linhas_nan = df_merged[df_merged['Região'].isna()]
print(linhas_nan['Bairro'].unique())

['Maracanã' 'Braz de Pina' 'Senador Vasconcelos' 'Lins de Vasconcelos'
 'Vila Kosmos' 'Parque Colúmbia' 'Recreio Dos Bandeirantes'
 'Vasco da Gama']


Os bairros que não estavam contidos na lista do DataFrame dados_bairros são: 'Maracanã', 'Braz de Pina', 'Senador Vasconcelos', 'Lins de Vasconcelos','Vila Kosmos', 'Parque Colúmbia' 'Recreio Dos Bandeirantes', 'Vasco da Gama'

Vamos procurar por um pedaço do nome do bairro e ver se não estão contidos ou se o problema é a grafia.

In [352]:
bairros_dados_bairros=dados_bairros['Bairro'].unique()

# Termos a serem procurados
termos_procurados = ["Mara", "Pina","Lin","Vila",'Sena','Parque','Recreio','Gama','Vasco']

# Procurar os termos na lista de strings
resultados = [s for s in bairros_dados_bairros for termo in termos_procurados if termo in s]

# Exibir os resultados
print("Resultados encontrados:")
for resultado in resultados:
    print(resultado)


Resultados encontrados:
Maracanâ
Vila Isabel
Vila Valqueire
Vila da Penha
Lins de Vasconcellos
Lins de Vasconcellos
Recreio dos Bandeirantes
Vila Cosmos
Vila Militar
Brás de Pina
Parque Anchieta
Parque Columbia
Senador Vasconcellos
Senador Vasconcellos
Senador Camará


Como podemos ver, "Maracanã", "Braz de Pina", "Senador Vasconcelos", "Lins de Vasconcelos", "Vila Kosmos","Parque Colúmbia"e "Recreio dos Bandeirantes" já estão na lista do DataFrame dados_bairros, o problema em questão é apenas a grafia que difere, somente "Vasco da Gama" não foi encontrado. Após uma breve pesquisa, descobrimos que as grafias corretas são "Maracanã", "Brás de Pina","Senador Vasconcelos", "Lins de Vasconcelos", "Vila Kosmos", "Parque Colúmbia" e "Recreio dos Bandeirantes". Ou seja, as únicas grafias que estão corretas no dados_bairros são as de "Brás de Pina" e "Recreio dos Bandeirantes". Logo precisamos saber se no df principal estão faltando ou se estão com a grafia errada. 

In [353]:
bairros_dados=dados['Bairro'].unique()

# Termos a serem procurados
termos_procurados = ["Mara", "Pina","Lin","Vila",'Sena','Parque','Recreio','Gama']

# Procurar os termos na lista de strings
resultados = [s for s in bairros_dados for termo in termos_procurados if termo in s]

# Exibir os resultados
print("Resultados encontrados:")
for resultado in resultados:
    print(resultado)

Resultados encontrados:
Recreio dos Bandeirantes
Parque Anchieta
Maracanã
Vila Isabel
Senador Camará
Braz de Pina
Senador Vasconcelos
Lins de Vasconcelos
Vila Valqueire
Vila Kosmos
Vila da Penha
Parque Colúmbia
Recreio Dos Bandeirantes
Vasco da Gama


No df dados os bairros "Brás de Pina" e "Recreio dos Bandeirantes" estão com a grafia errada. Após verificar isso, vamos substituir em ambos os dataframes pelos nomes corretos e refazer o merge. Como o bairro Vasco da Gama não está contido na nossa lista de bairros do dataframe dados_bairros, vamos precisar adicionar as informações manualmente. Este bairro pertence a Zona Central e a R.A São Cristóvão 

In [354]:

#Substituindo os nomes
substituicoes = {
    "Braz de Pina": "Brás de Pina",
    "Recreio Dos Bandeirantes": "Recreio dos Bandeirantes"
}
dados['Bairro'] = dados['Bairro'].replace(substituicoes)

substituicoes2 = {
    "Maracanâ": "Maracanã",
    "Senador Vasconcellos": "Senador Vasconcelos",
    "Lins de Vasconcellos": "Lins de Vasconcelos",
    "Parque Columbia": "Parque Colúmbia",
    "Vila Cosmos": "Vila Kosmos"
}
dados_bairros['Bairro'] = dados_bairros['Bairro'].replace(substituicoes2)
#Refazendo o merge
df_merged = pd.merge(left=dados, right=dados_bairros, how="left", on="Bairro")

#Acrescentando os dados do bairro Vasco da Gama
filtro = df_merged['Bairro'] == 'Vasco da Gama'
df_merged.loc[filtro, 'R.A'] = 'São Cristóvão'
df_merged.loc[filtro, 'Região'] = 'Zona Central'

Agora podemos verificar se possui apenas as 4 regiões e se todos os bairros foram classificas em uma região.

In [355]:
#Checkando se o dataframe possui as 4 entradas únicas na região.
print(len(df_merged['Região'].unique()))
print(df_merged['Região'].unique())

4
['Zona Sul' 'Zona Oeste' 'Zona Norte' 'Zona Central']


In [357]:
df_merged.head(10)

Preço                    Bairro     Categoria    Quartos Banheiros  \
0  R$ 870.000                  Botafogo  Apartamentos          1         1   
1  R$ 670.000               Jacarepaguá  Apartamentos          2         2   
2  R$ 690.000  Recreio dos Bandeirantes  Apartamentos          2         2   
3  R$ 500.000                Copacabana  Apartamentos          1         2   
4   R$ 22.000                Santa Cruz         Casas          2         1   
5  R$ 215.000           Parque Anchieta  Apartamentos          2         1   
6  R$ 390.000              Campo Grande         Casas          2         1   
7  R$ 400.000              Campo Grande         Casas          2         1   
8  R$ 650.000                Santíssimo         Casas  5 ou mais         4   
9  R$ 223.000                 Guaratiba         Casas          3         3   

      Nº              R.A      Região    IDS  
0   16.0         Botafogo    Zona Sul  743.0  
1  148.0      Jacarepaguá  Zona Oeste  476.0  
2   56.0  Barra da Tijuca  Zona Oeste  612.0  
3   14.0       Copacabana    Zona Sul  753.0  
4  147.0       Santa Cruz  Zona Oeste  476.0  
5  120.0         Anchieta  Zona Norte  540.0  
6  129.0     Campo Grande  Zona Oeste  518.0  
7  129.0     Campo Grande  Zona Oeste  518.0  
8  140.0     Campo Grande  Zona Oeste  491.0  
9  154.0        Guaratiba  Zona Oeste  433.0

In [358]:
#Número de linhas com valores faltantes por coluna
print(df_merged.isna().sum())


Preço          28
Bairro          0
Categoria       0
Quartos      1829
Banheiros    1839
Nº              2
R.A             0
Região          0
IDS             2
dtype: int64


In [359]:
#Renomeando os dados
dados = df_merged.copy()

Como podemos ver, ainda restam algumas linhas com valores faltantes. Elas poderiam ser removidas. Entretanto elas serão mantidas, pois dependendo do tipo de análise feita, elas podem ser interessantes.

Em um próximo trabalho essa base será utilizada para o ajuste de uma regressão.